File: hackerRank/java/tutorials/interview/BubbleSort.java
package interview;

import java.util.Arrays;
import java.util.Scanner;

/*
https://www.hackerrank.com/challenges/ctci-bubble-sort/problem
 */
public class BubbleSort {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int size = in.nextInt();
        int[] arr = new int[size];
        for (int i=0; i < size; i++) {
            arr[i] = in.nextInt();
        }
        in.close();

        System.out.println(Arrays.toString(arr));
        int swapCnt = 0;
        for (int i=0; i<size; i++) {
            for (int j=0; j<size-1; j++) {
                if(arr[j] > arr[j+1]) {
                   swapCnt++;
                   int temp = arr[j];
                   arr[j] = arr[j+1];
                   arr[j+1] = temp;
                }
            }
        }

        System.out.println(Arrays.toString(arr));
        System.out.println("Array is sorted in " + swapCnt + " swaps.");
        System.out.println("First Element: " + arr[0]);
        System.out.println("Last Element: " + arr[size - 1]);
    }
}

File: javascript/sort/bubbleSort.js
'use strict';

function sort(arr) {
  const len = arr.length;

  for (let i = len - 1; i > 0; i--) {
    let swap = 0;
    for (let j = 0; j < i; j++) {
      if (arr[j] > arr[j+1]) {
        let temp = arr[j];
        arr[j] = arr[j+1];
        arr[j+1] = temp;
        swap++;
      }
    }
    console.log(len - i, swap, arr);

    // 교환히 한번도 일어나지 않았다면, 이미 정렬이 완료된 것이다.
    if (swap === 0) {
      break;
    }
  }
  return arr;
}

// const arr = [22, 5, 11, 32, 120, 68, 70];
const arr = [1, 3, 9, 4, 7, 8, 6];
console.log('before', arr);
console.log('after', sort(arr));
